### ISSUE TITLE
shared_ptr.unique() prevents C++20 / Native windows build

### ISSUE THREAD

**Issue by LeeReid1** [MRtrix3/mrtrix3#2894]
Hi guys,

I need to preface this by highlighting C++ is not my mainstay and I've not your depth of knowledge on this codebase, so am a little out of my depth here.

I've locally branched the dev arm to put together a native windows build (i.e. MSVC). MSVC cannot resolve some of the more complex SFINAE code in 2-3 files, and bugs-out on a supporting library, but this is resolvable by switching to C++20 concepts. Suffice to say it's an absolute blackhole time-wise to find a work-around, if staying in C++17. 
My WIP currently should still compile as normal on linux (C++17) and is almost there on Windows with C++20 for most executables. 

However, the compilation fails due to `shared_ptr.unique()` in `image.h`. This method is deprecated in C++17 and eliminated in C++20 as [it's not threadsafe](https://stackoverflow.com/questions/41142315/why-is-stdshared-ptrunique-deprecated) and checks are likely avoidable by being careful with code architecture. 

I can't help but wonder about the code using it, but my knowledge of how mrtrix is put together is not deep enough to fix this without potentially causing an issue elsewhere. Hence, I'm seeking your opinions on solutions.

The primary use is in the destructor of `image.h`, where it seems to be used to check if `this` is the only user of the buffer. What follows seems to be unbuffering back to some original location(?): 

```
template <typename ValueType> Image<ValueType>::~Image() {
  if (buffer.unique()) { 
    // was image preloaded and read/write? If so,need to write back:
    if (buffer->get_io()) {
      if (buffer->get_io()->is_image_readwrite() && buffer->data_buffer) {
        auto data_buffer = std::move(buffer->data_buffer);
        TmpImage<ValueType> src = {
            *buffer, data_buffer.get(), std::vector<ssize_t>(ndim(), 0), strides, Stride::offset(*this)};
        Image<ValueType> dest(buffer);
        threaded_copy_with_progress_message("writing back direct IO buffer for \"" + name() + "\"", src, dest);
      }
    }
  }
}
```

I find it odd that an image may not be the owner or sole consumer of its own buffer in any read/write scenario. Other code suggests that, actually, this situation should never occur:

```
template <typename ValueType> Image<ValueType> Image<ValueType>::with_direct_io(Stride::List with_strides) {
  if (buffer->data_buffer)
    throw Exception("FIXME: don't invoke 'with_direct_io()' on images already using direct IO!");
  if (!buffer->get_io())
    throw Exception("FIXME: don't invoke 'with_direct_io()' on non-validated images!");
  if (!buffer.unique())
    throw Exception("FIXME: don't invoke 'with_direct_io()' on images if other copies exist!");
  
```

That suggests to me that the `buffer.unique()` check in `~` could simply be removed, and perhaps this safety check in `with_direct_io` could be avoided by use/check-for a `unique_ptr` or similar. If that's enough, I'm happy.

I believe there's another workaround I've not attempted, which would be something like so:
```
bool owns_buffer(){
  std::weak_ptr<int> wp = buffer;
  if (auto locked = wp.lock()) {
    return true;
  }
  return false;
}
```

That said, I can't help but wonder if there's a deeper issue. OOP principles suggest an object should only create/expose its data in a 'safe' way, eliminating the need for such checks. So, if Image1 opens a file and needs to make a buffer, noone else should be able to take control of that buffer, and if it's exposed it's up to anyone reading it to stop doing so before the owner is destroyed. If the dependency injection pattern is relied upon, the creator of resources would be responsible for their clean-up/finalisation, not the consumer of them. So, if Image1 is provided a buffer to read/write to, it is not responsible for what happens to that buffer after it's finished its work.

For example:
* Image opens file x.mif
* Image creates a buffer, or some translation layer, to be able to read/write to it
* Image cleans up resources when done

Or with DI:
* Image1 accepts a `dataProvider` agnostically
* Image1 reads/writes to `dataProvider` as needed
* `dataProvider` takes responsibility to convert that into appropriate I/O operations (to disk/memory/etc). 
* Either the `dataProvider` handles clean up in its destructor, or the creator of Image1 explicitly cleans it up when Image1 is finished being used.

The current code looks like it's trying to use DI, but then on its destructor take responsibility for figuring out if it's been abandoned and needs to be a good citizen and complete some IO operation that presumably was intended by its creator(?).

I've probably misinterpreted something due to lack of context but any help you could offer eliminating these `unique()` calls would be really appreciated. Will offer the native build back as a PR if I can get it over the line.

Thanks

**Comment by daljit46** [comment]
Hi, you can try to replace `unique()` with `use_count() == 1` (see [here](https://en.cppreference.com/w/cpp/memory/shared_ptr/use_count). This should at least make that compile. Whether those checks can be removed needs more thinking, but I agree that we should try to get rid of it.
Getting a native MSVC build would be nice and something I'd like to make viable just to use all the Visual Studio tools, but I haven't attempted that yet as currently the codebase is quite Unix-centric.

**Comment by LeeReid1** [comment]
Can confirm that works! Many thanks. Given my goal is just to get it working I won't attempt any kind of reworking of the architecture leading to the problem. Up to you whether you'd like to close this issue or leave it open as something to review at a later stage.

**Comment by bjeurissen** [comment]
This is now emitting a lot of warnings when compiling with clang 16 (e.g. on macOS 15) 
```
/Users/ben/mrtrix3_dev/core/image.h:363:15: warning: 'unique' is deprecated [-Wdeprecated-declarations]
  363 |   if (!buffer.unique())
      |               ^
/Users/ben/mrtrix3_dev/src/gui/mrview/tool/fixel/directory.cpp:55:95: note: in instantiation of member function 'MR::Image<float>::with_direct_io' requested here
   55 |       MR::Fixel::find_directions_header(Path::dirname(fixel_data->name())).get_image<float>().with_direct_io();
      |                                                                                               ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk/usr/include/c++/v1/__memory/shared_ptr.h:730:3: note: 'unique' has been explicitly marked deprecated here
  730 |   _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_HIDE_FROM_ABI bool unique() const _NOEXCEPT { return use_count() == 1; }
      |   ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk/usr/include/c++/v1/__config:1022:41: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
 1022 | #    define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
      |                                         ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk/usr/include/c++/v1/__config:995:49: note: expanded from macro '_LIBCPP_DEPRECATED'
  995 | #      define _LIBCPP_DEPRECATED __attribute__((__deprecated__))
      |                                                 ^
```

**Comment by daljit97** [comment]
@jdtournier would you mind commenting (if you can remember :) ) on the reasoning behind using `shared_ptr` and `unique()`? I think we should try to get rid of it, but it'd be good to have a good grasp on why the code was written in this manner first.

**Comment by daljit46** [comment]
@jdtournier would you mind commenting on the reasoning behind using a `shared_ptr` and `unique()` (if you can still remember :))? I think we should get rid of it, but it'd be good to get a grasp on why the code was written in this manner first.

**Comment by jdtournier** [comment]
Ok, that code takes me back a bit... It's going to take be a while to piece all my reasoning back together again, but my first thought here is that the delayed write-back that is dealing with really should be done in the Buffer destructor - not the Image destructor. This delayed write-back was bolted on later to avoid the use of memory-mapping on network or otherwise slow filesystems, which caused issues otherwise. This might explain why the implementation might not have been thought through all that properly...

**Comment by Lestropie** [comment]
Bumping as I'm trying to fix compilation issues that only appear on CI and not on my own machine, and this warning is everywhere.

I'm not sure that there's going to be getting around the write-back-in-destructor design. The point of the design is that `MR::Image` could be copy-constructed in all sorts of different places, there's no singular "owner" of the data, and so in the case of delayed write-back it has to be the last one to leave that closes the door. The fundamental issue that has led to the deprecation in the standard is that `std::shared_ptr<>::unique()` is *not thread-safe*. Which is a pretty bad oversight, but whatever.

So I think the solution here is to define a class that derives from `std::shared_ptr<>`, which supersedes the use counter with a thread-safe version (`std::atomic<int>` with `std::memory_order::seq_cst` should suffice? Otherwise `std::mutex`), and can therefore also provide a safe `.unique()` member function.

**Comment by daljit46** [comment]
> So I think the solution here is to define a class that derives from std::shared_ptr<>, which supersedes the use counter with a thread-safe version (std::atomic<int> with std::memory_order::seq_cst should suffice? Otherwise std::mutex), and can therefore also provide a safe .unique() member function.

I don't think this is the direction we should take in. Firstly, we should try to avoid inheriting from `std::shared_ptr`. Most STL types aren't built for this purpose and you can run into very subtle problems if you inherit from them. Secondly, `std::shared_ptr` has a hidden control block where its count is stored and this is not accessible from the "object" you hold in a variable. Thus its ownership mechanism is opaque and cannot really be replaced. Even if we add a `std::atomic<int32_t>` as a data member of the derived class, we'll effectively duplicate the existing ownership of `shared_ptr` (requiring them to be in lockstep all the time).

I think in line with what @jdtournier said, what we should do here is to move ownership of the data into the `Buffer` class (it should also be responsible for the delayed-write-back). The `Image` class should be solely be treated as a view.
